SimPEG/utils/io_utils/io_utils_electromagnetics.py

```diff
@@ -1,35 +1,208 @@
+    
+    # Prevent circular import
+    from ...electromagnetics.static import resistivity as dc
+    from ...data import Data
+    
+    # Load file headers
+    FID = open(file_name, 'r')
+    file_headers = FID.readline()
+    FID.close()
+    file_headers = file_headers.split()
+
+    # Find indices of columns being loaded
+    out_headers = a_headers + b_headers + m_headers + n_headers
+    
+    has_data = False
+    has_uncert = False
+    has_dict = False
+    non_location_columns = 0
+    if data_header is not None:
+        out_headers.append(data_header)
+        has_data = True
+        non_location_columns += 1
+    if uncertainties_header is not None:
+        out_headers.append(uncertainties_header)
+        has_uncert = True
+        non_location_columns += 1
+    if dict_headers is not None:
+        out_headers += dict_headers
+        has_dict = True
+        non_location_columns += len(dict_headers)
+    
+    col_indices = []
+    for h in out_headers:
+        col_indices.append(file_headers.index(h))
+    
+    # Load specified columns of data
+    data_array = np.loadtxt(
+        file_name, comments='!', skiprows=1, usecols=col_indices
+    )
+    n_rows = np.shape(data_array)[0]
+    num_location_columns = np.shape(data_array)[1] - non_location_columns
+    
+    # Extract columns for electrode locations
+    if num_location_columns == 12:
+        a_cols = [0, 1, 2]
+        b_cols = [3, 4, 5]
+        m_cols = [6, 7, 8]
+        n_cols = [9, 10, 11]
+    # 2D survey data or surface 3D data
+    elif num_location_columns == 8:
+        a_cols = [0, 1]
+        b_cols = [2, 3]
+        m_cols = [4, 5]
+        n_cols = [6, 7]
+    # 2D surface data
+    else:
+        a_cols = [1]
+        b_cols = [2]
+        m_cols = [3]
+        n_cols = [4]
+    
+    # Extract electrode locations
+    if is_surface_data:
+        dummy_elevation = 9999  # Taller than mount Everest
+        a_locations = np.c_[data_array[:, a_cols], dummy_elevation*np.ones(n_rows)]
+        b_locations = np.c_[data_array[:, b_cols], dummy_elevation*np.ones(n_rows)]
+        m_locations = np.c_[data_array[:, m_cols], dummy_elevation*np.ones(n_rows)]
+        n_locations = np.c_[data_array[:, n_cols], dummy_elevation*np.ones(n_rows)]
+        warnings.warn(
+            "Loaded data are in surface format. Elevations automatically set to 9999 m. "
+            "Use the project_to_discretized_topography method of the survey to project "
+            "electrode locations to the discretized surface."
+        )
+    else:
+        a_locations = data_array[:, a_cols]
+        b_locations = data_array[:, b_cols]
+        m_locations = data_array[:, m_cols]
+        n_locations = data_array[:, n_cols]
+    
+    # Set up keeping track of sorting of rows and unique sources
+    k = np.arange(0, n_rows)
+    out_indices = []
+    unique_ab, ab_index = np.unique(np.c_[a_locations, b_locations], axis=0, return_index=True)
+    ab_index = np.sort(ab_index)
```

Comment:
I would like to see our generating functions that read in a file, put it in the most memory efficient format for SimPEG. I.E. make a list of all unique sources, and all unique receivers. Create objects for each of them. Then go through and attach the receivers to the sources. (This stops repeated definitions of projection matrices when the receivers are in the same location).

Reply:
Don't we have a function that is survey_from_abmn, that could be called here?